### ISSUE TITLE
Write our take on `Optional` vs. null-augmented types

### ISSUE THREAD

**Issue by cpovirk** [jspecify/jspecify#591]
Examples of existing discussions:

- https://homes.cs.washington.edu/~mernst/advice/nothing-is-better-than-optional.html
- https://www.reddit.com/r/java/comments/13rv1t2/why_exactly_does_brian_goetz_not_recommend_the/
- https://www.reddit.com/r/java/comments/1epg4cf/comment/lhl0ay3/

Points to possibly hit from those and other discussions:
- Much more static analysis exists to prevent dereferencing of nullable types than unwrapping of empty `Optional` instances, and that will only grow if [the draft JEP](https://bugs.openjdk.org/browse/JDK-8303099) progresses.
- `!` and `?` will be a lot nicer than `Optional` if the draft JEP progresses, too.
- Kotlin already handles nullable types well.
- Many more existing APIs use `null` than `Optional`, so there's no fully avoiding `null` (or `Optional`, of course, but to a lesser extent).
- If a method returns `Optional<Foo>`, you can't override it to return `Foo`. Similarly, if a method accepts `Optional<Foo>`, you have to wrap your `Foo` to pass it in. (This is (I think) another angle on "not composable" from rzwitserloot's post.)
- Similarly, we've found that it's very easy to accidentally write `assertThat(someOptional).isEqualTo(someValue)` instead of `assertThat(someOptional).hasValue(someValue)` (and similarly for other comparisons).
- `Optional` isn't `Serializable` (not that we advise using serialization if you can get away with it).

My "oversimplified" summary [last month](https://github.com/jspecify/jspecify/issues/544#issuecomment-2214782316) was:

> I tend to think of `Optional` as a tool to use in APIs whose callers you can't count on to use nullness checkers.

**Comment by kevinb9n** [comment]
Good idea. The idea that Optional is some kind of substitute for null-aware types comes up _constantly_.

One crux of it for me is that I can change a variable from `Foo` to `Foo?` and I only have to fix the usages of that variable that were actually _depending_ on its nullness. The rest are fine. Changing to `Optional<Foo>` is _much_ more invasive. Of course, I also lose all the benefits nullness checking!

(Or, if anyone _has_ created static analyzers that try to track whether an Optional can be absent or not and warn appropriately, imho that's even sadder, just wasting effort to reinvent nullability all over again!)

If `Optional` is good for anything, at most it's for cases where you're never going to actually _need_ the thing to be present; you just want to cover both cases in your code. If you like the _API_ that it gives you for doing that, then cool. It was never a general-purpose replacement for null.

**Comment by agentgt** [comment]
Personally I have made all those points numerous times on reddit and surprisingly there seems to always be folks that just I think refuse to admit the faults (perhaps cognitive dissonance or they have already invested into optional).

I particularly like the switching to future nonnull/null types argument that Kevin uses but still people will just argue back and forth on it that somehow valhalla will fix it.

I guess each have their own argument that are in the annotation camp but mine is more meta and along the above that you cannot express `Optional` contract with `Optional`.

For example Optional.map takes `Function<T,R extends @Nullable Object>` and not `Function<T, R>`  (the map function should do `Objects.requireNonNull` before it hands off the value). That is if it were designed to remove `null` it should not do what it does with `map` and actually encourage `null` being used.

Anyway what I'm getting at is I'm just not sure it is worth expending a lot of time on. If Kevin just says it is going to be difficult to switch later if use Optional is IMO good enough.

**Comment by cpovirk** [comment]
A thing that I think you're both suggesting but that didn't immediately register with me is one of our usual points: The Java language is investigating adding support for nullness. So we can expect the ecosystem to shift further in that direction, and we should also notice that migrating from `@Nullable Foo` to `Foo?` will be a lot easier than migrating from `Optional<Foo>` to `Foo` (including that the former is binary-compatible).

(And yes, this issue is very much prompted by various Reddit discussions :))
